File: too_many_cooks_data/test/plans_test.dart
/// Tests for agent plan operations.
library;

import 'dart:js_interop';

import 'package:dart_node_core/dart_node_core.dart';
import 'package:nadz/nadz.dart';
import 'package:test/test.dart';
import 'package:too_many_cooks_data/too_many_cooks_data.dart';

extension type _Fs(JSObject _) implements JSObject {
  external void unlinkSync(String path);
  external bool existsSync(String path);
}

final _Fs _fs = _Fs(requireModule('fs') as JSObject);

void _deleteIfExists(String path) {
  try {
    if (_fs.existsSync(path)) {
      _fs.unlinkSync(path);
    }
  } on Object catch (_) {}
}

void main() {
  const testDbPath = '.test_plans.db';
  TooManyCooksDb? db;
  var agentName = '';
  var agentKey = '';

  setUp(() {
    _deleteIfExists(testDbPath);
    final config = createDataConfig(dbPath: testDbPath);
    final result = createDb(config);
    expect(result, isA<Success<TooManyCooksDb, String>>());
    db = (result as Success<TooManyCooksDb, String>).value;

    // Register test agent
    final regResult = db!.register('plan-agent');
    final reg = (regResult as Success<AgentRegistration, DbError>).value;
    agentName = reg.agentName;
    agentKey = reg.agentKey;
  });

  tearDown(() {
    db?.close();
    _deleteIfExists(testDbPath);
  });

  test('updatePlan creates new plan', () {
    final result = db!.updatePlan(
      agentName,
      agentKey,
      'Fix all bugs',
      'Reading codebase',
    );
    expect(result, isA<Success<void, DbError>>());
  });

  test('updatePlan updates existing plan', () {
    db!.updatePlan(agentName, agentKey, 'Goal 1', 'Task 1');

    final result = db!.updatePlan(agentName, agentKey, 'Goal 2', 'Task 2');
    expect(result, isA<Success<void, DbError>>());

    final getPlan = db!.getPlan(agentName);
    final plan = (getPlan as Success<AgentPlan?, DbError>).value!;
    expect(plan.goal, 'Goal 2');
    expect(plan.currentTask, 'Task 2');
  });

  test('updatePlan fails with invalid credentials', () {
    final result = db!.updatePlan(agentName, 'wrong-key', 'Goal', 'Task');
    expect(result, isA<Error<void, DbError>>());
    final error = (result as Error<void, DbError>).error;
    expect(error.code, errUnauthorized);
  });

  test('updatePlan fails for goal exceeding max length', () {
    final longGoal = 'x' * 101; // Default max is 100
    final result = db!.updatePlan(agentName, agentKey, longGoal, 'Task');
    expect(result, isA<Error<void, DbError>>());
    final error = (result as Error<void, DbError>).error;
    expect(error.code, errValidation);
    expect(error.message, contains('100'));
  });

  test('updatePlan fails for task exceeding max length', () {
    final longTask = 'x' * 101;
    final result = db!.updatePlan(agentName, agentKey, 'Goal', longTask);
    expect(result, isA<Error<void, DbError>>());
    final error = (result as Error<void, DbError>).error;
    expect(error.code, errValidation);
  });

  test('getPlan returns plan for agent', () {
    db!.updatePlan(agentName, agentKey, 'My Goal', 'Current Task');

    final result = db!.getPlan(agentName);
    expect(result, isA<Success<AgentPlan?, DbError>>());
    final plan = (result as Success<AgentPlan?, DbError>).value;
    expect(plan, isNotNull);
    expect(plan!.agentName, agentName);
    expect(plan.goal, 'My Goal');
    expect(plan.currentTask, 'Current Task');
    expect(plan.updatedAt, greaterThan(0));
  });

  test('getPlan returns null for agent without plan', () {
    // Register agent without setting plan
    final reg2 = db!.register('no-plan-agent');
    final agent2 = (reg2 as Success<AgentRegistration, DbError>).value;

    final result = db!.getPlan(agent2.agentName);
    expect(result, isA<Success<AgentPlan?, DbError>>());
    final plan = (result as Success<AgentPlan?, DbError>).value;
    expect(plan, isNull);
  });

  test('listPlans returns all plans', () {
    db!.updatePlan(agentName, agentKey, 'Goal 1', 'Task 1');

    // Register second agent with plan
    final reg2 = db!.register('plan-agent-2');
    final agent2 = (reg2 as Success<AgentRegistration, DbError>).value;
    db!.updatePlan(agent2.agentName, agent2.agentKey, 'Goal 2', 'Task 2');

    final result = db!.listPlans();
    expect(result, isA<Success<List<AgentPlan>, DbError>>());
    final plans = (result as Success<List<AgentPlan>, DbError>).value;
    expect(plans.length, 2);
    expect(plans.map((p) => p.goal).toSet(), {'Goal 1', 'Goal 2'});
  });

  test('plan updatedAt changes on update', () {
    db!.updatePlan(agentName, agentKey, 'Goal', 'Task 1');
    final getPlan1 = db!.getPlan(agentName);
    final plan1 = (getPlan1 as Success<AgentPlan?, DbError>).value;

    db!.updatePlan(agentName, agentKey, 'Goal', 'Task 2');
    final getPlan2 = db!.getPlan(agentName);
    final plan2 = (getPlan2 as Success<AgentPlan?, DbError>).value;

    expect(plan2!.updatedAt, greaterThanOrEqualTo(plan1!.updatedAt));
  });
}

File: too-many-cooks/packages/too-many-cooks/test/plans_test.ts
/// Tests for agent plan operations.

import { describe, it, beforeEach, afterEach } from "node:test";
import assert from "node:assert";
import fs from "node:fs";
import {
  type TooManyCooksDb,
  createDataConfig,
  ERR_UNAUTHORIZED,
  ERR_VALIDATION,
} from "@too-many-cooks/core";
import { createDb } from "../src/db-sqlite.js";

const TEST_DB_PATH = ".test_plans.db";

const deleteIfExists = (path: string): void => {
  try {
    if (fs.existsSync(path)) {
      fs.unlinkSync(path);
    }
  } catch {
    // ignore
  }
};

describe("plans", () => {
  let db: TooManyCooksDb | undefined;
  let agentName = "";
  let agentKey = "";

  beforeEach(async () => {
    deleteIfExists(TEST_DB_PATH);
    const config = createDataConfig({ dbPath: TEST_DB_PATH });
    const result = createDb(config);
    assert.strictEqual(result.ok, true);
    if (!result.ok) {throw new Error("expected ok");}
    db = result.value;

    // Register test agent
    const regResult = await db.register("plan-agent");
    if (!regResult.ok) {throw new Error("expected ok");}
    const reg = regResult.value;
    agentName = reg.agentName;
    agentKey = reg.agentKey;
  });

  afterEach(async () => {
    await db?.close();
    deleteIfExists(TEST_DB_PATH);
  });

  it("updatePlan creates new plan", async () => {
    assert.notStrictEqual(db, undefined);
    if (!db) {throw new Error("expected db");}
    const result = await db.updatePlan(
      agentName,
      agentKey,
      "Fix all bugs",
      "Reading codebase",
    );
    assert.strictEqual(result.ok, true);
  });

  it("updatePlan updates existing plan", async () => {
    assert.notStrictEqual(db, undefined);
    if (!db) {throw new Error("expected db");}
    await db.updatePlan(agentName, agentKey, "Goal 1", "Task 1");

    const result = await db.updatePlan(agentName, agentKey, "Goal 2", "Task 2");
    assert.strictEqual(result.ok, true);

    const getPlan = await db.getPlan(agentName);
    if (!getPlan.ok) {throw new Error("expected ok");}
    const plan = getPlan.value!;
    assert.strictEqual(plan.goal, "Goal 2");
    assert.strictEqual(plan.currentTask, "Task 2");
  });

  it("updatePlan fails with invalid credentials", async () => {
    assert.notStrictEqual(db, undefined);
    if (!db) {throw new Error("expected db");}
    const result = await db.updatePlan(agentName, "wrong-key", "Goal", "Task");
    assert.strictEqual(result.ok, false);
    if (result.ok) {throw new Error("expected error");}
    assert.strictEqual(result.error.code, ERR_UNAUTHORIZED);
  });

  it("updatePlan fails for goal exceeding max length", async () => {
    assert.notStrictEqual(db, undefined);
    if (!db) {throw new Error("expected db");}
    const longGoal = "x".repeat(101); // Default max is 100
    const result = await db.updatePlan(agentName, agentKey, longGoal, "Task");
    assert.strictEqual(result.ok, false);
    if (result.ok) {throw new Error("expected error");}
    assert.strictEqual(result.error.code, ERR_VALIDATION);
    assert.ok(result.error.message.includes("100"));
  });

  it("updatePlan fails for task exceeding max length", async () => {
    assert.notStrictEqual(db, undefined);
    if (!db) {throw new Error("expected db");}
    const longTask = "x".repeat(101);
    const result = await db.updatePlan(agentName, agentKey, "Goal", longTask);
    assert.strictEqual(result.ok, false);
    if (result.ok) {throw new Error("expected error");}
    assert.strictEqual(result.error.code, ERR_VALIDATION);
  });

  it("getPlan returns plan for agent", async () => {
    assert.notStrictEqual(db, undefined);
    if (!db) {throw new Error("expected db");}
    await db.updatePlan(agentName, agentKey, "My Goal", "Current Task");

    const result = await db.getPlan(agentName);
    assert.strictEqual(result.ok, true);
    if (!result.ok) {throw new Error("expected ok");}
    const plan = result.value;
    assert.notStrictEqual(plan, undefined);
    assert.strictEqual(plan!.agentName, agentName);
    assert.strictEqual(plan!.goal, "My Goal");
    assert.strictEqual(plan!.currentTask, "Current Task");
    assert.ok(plan!.updatedAt > 0);
  });

  it("getPlan returns null for agent without plan", async () => {
    assert.notStrictEqual(db, undefined);
    if (!db) {throw new Error("expected db");}
    // Register agent without setting plan
    const reg2 = await db.register("no-plan-agent");
    if (!reg2.ok) {throw new Error("expected ok");}
    const agent2 = reg2.value;

    const result = await db.getPlan(agent2.agentName);
    assert.strictEqual(result.ok, true);
    if (!result.ok) {throw new Error("expected ok");}
    const plan = result.value;
    assert.strictEqual(plan, null);
  });

  it("listPlans returns all plans", async () => {
    assert.notStrictEqual(db, undefined);
    if (!db) {throw new Error("expected db");}
    await db.updatePlan(agentName, agentKey, "Goal 1", "Task 1");

    // Register second agent with plan
    const reg2 = await db.register("plan-agent-2");
    if (!reg2.ok) {throw new Error("expected ok");}
    const agent2 = reg2.value;
    await db.updatePlan(agent2.agentName, agent2.agentKey, "Goal 2", "Task 2");

    const result = await db.listPlans();
    assert.strictEqual(result.ok, true);
    if (!result.ok) {throw new Error("expected ok");}
    const plans = result.value;
    assert.strictEqual(plans.length, 2);
    assert.deepStrictEqual(
      new Set(plans.map((p) => p.goal)),
      new Set(["Goal 1", "Goal 2"]),
    );
  });

  it("plan updatedAt changes on update", async () => {
    assert.notStrictEqual(db, undefined);
    if (!db) {throw new Error("expected db");}
    await db.updatePlan(agentName, agentKey, "Goal", "Task 1");
    const getPlan1 = await db.getPlan(agentName);
    if (!getPlan1.ok) {throw new Error("expected ok");}
    const plan1 = getPlan1.value;

    await db.updatePlan(agentName, agentKey, "Goal", "Task 2");
    const getPlan2 = await db.getPlan(agentName);
    if (!getPlan2.ok) {throw new Error("expected ok");}
    const plan2 = getPlan2.value;

    assert.ok(plan2!.updatedAt >= plan1!.updatedAt);
  });
});
